lib/Conversion/BGVToPoly/BGVToPoly.cpp

```diff
@@ -97,6 +97,60 @@ struct ConvertNegate : public OpConversionPattern<Negate> {
   }
 };
 
+struct ConvertMul : public OpConversionPattern<MulOp> {
+  ConvertMul(mlir::MLIRContext *context)
+      : OpConversionPattern<MulOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      MulOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto x = adaptor.getX();
+    auto xT = dyn_cast<RankedTensorType>(x.getType());
+    auto y = adaptor.getY();
+    auto yT = dyn_cast<RankedTensorType>(y.getType());
+
+    if (xT.getNumElements() != 2 || yT.getNumElements() != 2) {
+      op.emitError() << "`bgv.mul` expects ciphertext as two polynomials, got "
+                     << xT.getNumElements() << " and " << yT.getNumElements();
+      return failure();
+    }
+
+    if (xT.getElementType() != yT.getElementType()) {
+      op->emitOpError() << "`bgv.mul` expects operands of the same type";
+      return failure();
+    }
+
+    auto loc = op.getLoc();
```

Comment:
You may want to use this to simplify the rewriter's loc! https://mlir.llvm.org/doxygen/classmlir_1_1ImplicitLocOpBuilder.html